### PR TITLE
linux: implement fallocate(mode=0) compatibility

### PR DESCRIPTION
**Pull request by adilger**
Implement semi-compatible functionality for mode=0 (preallocation)
and mode=FALLOC_FL_KEEP_SIZE (preallocation beyond EOF) for ZPL.

Since ZFS does COW and snapshows, preallocating blocks for a file
cannot guarantee that writes to the file will not run out of space.
Instead, make a best-effort attempt to check that at least enough
space is currently available in the pool (12% margin), then create
a sparse file of the requested size and continue on with life.

The parameter of zfs_statvfs() is changed to take an inode instead
of a dentry, since no dentry is available in zfs_fallocate_common().

Signed-off-by: Andreas Dilger <adilger@dilger.ca>

### Motivation and Context

Allow fallocate(mode=0/2) to be used in a reasonable manner for ZFS, even though it can never exactly match the semantics of non-COW filesystems.

This partially addresses issue #326 but does not implement all of the available modes.

### Description

Implement semi-compatible functionality for mode=0 (preallocation)
and mode=FALLOC_FL_KEEP_SIZE (preallocation beyond EOF) for ZPL.

Since ZFS does COW and snapshows, preallocating blocks for a file
cannot guarantee that writes to the file will not run out of space.
Instead, make a best-effort attempt to check that at least enough
space is currently available in the pool (12% margin), then create
a sparse file of the requested size and continue on with life.

The parameter of zfs_statvfs() is changed to take an inode instead
of a dentry, since no dentry is available in zfs_fallocate_common().

### How Has This Been Tested?

Basic build test.  Has not had any functional testing, just for review and improvement.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ahrens** [comment]
How did you decide on the 12% margin? Why not allow it to succeed as long as there's enough free space (with no margin)?  If you use fallocate() on a file that's already filled in, what is that supposed to do?  Would we still want to check for free space, if there are snapshots?

File: module/os/linux/zfs/zpl_file.c [hunk 1]
```diff
@@ -743,21 +746,46 @@ zpl_fallocate_common(struct inode *ip, int mode, loff_t offset, loff_t len)
...
+		if (offset + len > olen)
+			len = olen - offset;
+		bf.l_type = F_WRLCK;
+		bf.l_whence = SEEK_SET;
+		bf.l_start = offset;
+		bf.l_len = len;
+		bf.l_pid = 0;
+
+		error = -zfs_space(ITOZ(ip), F_FREESP, &bf, O_RDWR, offset, cr);
+	} else if ((mode & ~FALLOC_FL_KEEP_SIZE) == 0) {
```
Comment (rlaager) [inline module/os/linux/zfs/zpl_file.c:766]: I'm having trouble wrapping my head around this. You've already ensured (via the check that returns EOPNOTSUPP) that nothing besides PUNCH_HOLE or KEEP_SIZE is set. This is the else arm, so we know that PUNCH_HOLE is not set. So at this point, KEEP_SIZE may or may not be set, but that is it. It looks like this condition is always true. I might be missing something, though, as I don't work with bitwise flags very often.

File: module/os/linux/zfs/zpl_file.c [hunk 2]
```diff
@@ -743,21 +746,46 @@ zpl_fallocate_common(struct inode *ip, int mode, loff_t offset, loff_t len)
...
-	bf.l_type = F_WRLCK;
-	bf.l_whence = SEEK_SET;
-	bf.l_start = offset;
-	bf.l_len = len;
-	bf.l_pid = 0;
-
 	crhold(cr);
 	cookie = spl_fstrans_mark();
-	error = -zfs_space(ITOZ(ip), F_FREESP, &bf, O_RDWR, offset, cr);
+	if (mode & FALLOC_FL_PUNCH_HOLE) {
```
Comment (rlaager) [inline module/os/linux/zfs/zpl_file.c:751]: The fallocate(2) man page says that PUNCH_HOLE must have KEEP_SIZE specified. Do we want to enforce that? We were previously.

**Comment by adilger** [comment]
> How did you decide on the 12% margin? Why not allow it to succeed as long as there's enough
> free space (with no margin)?

The 12% margin is totally arbitrary.  It is enough for indirect blocks and other file metadata (AFAIK that is about 3% for ZFS, along with minimum free space requirements just to use the filesystem), and considerations like if the file is not overwritten strictly sequentially or if there are some other "casual" writes to the filesystem.  I don't think it makes sense to return success for a 1MB fallocate() when there is exactly 1048576 bytes of free space available, as it seems unlikely that this will succeed.  At the same time, it should prevent cases like `fallocate(1TB)` on a 100GB filesystem succeeding and misleading the users.  A more sophisticated approach might be to have a pool/dataset parameter (e.g. `zpl_fallocate_ratio=N`) that adjusts the ratio of free space needed, and also allows setting it to 0 disables fake fallocate() and returns to returning to the old behavior:

        if (zpl_fallocate_ratio == 0)
                error = -EOPNOTSUP;
        else if (len * zpl_fallocate_ratio / 100 > availbytes)
                error = -ENOSPC;

In theory it would even be possible to set `zpl_fallocate_ratio` < 100 to account for compression.

> If you use fallocate() on a file that's already filled in, what is that supposed to do?
> Would we still want to check for free space, if there are snapshots?

I initially thought that having previously-allocated blocks might complicate matters, but in further reflection I don't think it does.  If the file is fully or partly allocated with snapshots, we _still_ need an equal amount of space available to write out the file again, so the fallocate() call essentially becomes a check for "is there currently enough free space for overwriting this file" rather than a reservation, but better than nothing.

File: module/os/linux/zfs/zpl_file.c (hunk 2 above)
Comment (adilger) [reply module/os/linux/zfs/zpl_file.c:751]: This is checked in the VFS already:

    int vfs_fallocate(struct file *file, int mode, loff_t offset, loff_t len)
    {
        :
        :
        /* Punch hole must have keep size set */
        if ((mode & FALLOC_FL_PUNCH_HOLE) &&
            !(mode & FALLOC_FL_KEEP_SIZE))
                return -EOPNOTSUPP;

File: module/os/linux/zfs/zpl_file.c (hunk 1 above)
Comment (adilger) [reply module/os/linux/zfs/zpl_file.c:766]: Correct. If you look at your ABCD table, the if block covers C/D when PUNCH_HOLE is set, so this _has_ to be the A/B case.  It could have just been "else" right now, but I thought it was more forward looking to check for the `mode=0/2` case explicitly (file preallocation, without/with size change) to facilitate addition of more modes later in separate else blocks (or possibly a`switch`).  Even without the `-EOPNOTSUPP` check, we would know that this is `mode=0/2` since the `~KEEP_SIZE` mask would pass any other bits through and the result would not `== 0` in that case.

Comment (adilger) [reply module/os/linux/zfs/zpl_file.c:766]: FYI, for Boolean logic we learned [Karnaugh Maps](https://en.wikipedia.org/wiki/Karnaugh_map) in engineering to simplify the logic, which can be very helpful when working with multiple variables and trying to write a simplified expression to handle the cases.

**Comment by kithrup** [comment]
I haven't looked at the code, but from the description it sounds like it's only a little different from ftruncate?

**Comment by adilger** [comment]
@kithrup, one significant difference is that this will fail if the fallocate() requests more space than there is free in the filesystem, which ftruncate() will not. Although it doesn't actually reserve the space, this provides an easy heads-up for applications that writing so much data to the file would (likely) fail, rather than writing GB of zeroes or data and hitting ENOSPC.

**Comment by kithrup** [comment]
As I said, only a little bit different from ftruncate.

Without actually reserving the space, it's a race condition, and you may as well just make a user-mode wrapper for fallocate() that calls fstatfs and ftruncate -- there is no semantic difference at that point.

**Comment by rlaager** [comment]
The main point here is that real applications already use ftruncate() and with ZFS not implementing these modes, the applications fall back to even worse behaviors, like writing GBs of zeros.

**Comment by arshad512** [comment]
Hi @rlaager , the main purpose of fallocate() is not to do large writes for space reservation. There is no plan to do it. Here if the modes are not supported, EOPNOTSUPP is returned immediately.

File: module/os/linux/zfs/zpl_file.c [hunk 3]
```diff
@@ -743,21 +746,46 @@ zpl_fallocate_common(struct inode *ip, int mode, loff_t offset, loff_t len)
...
+		 */
+		error = zfs_statvfs(ip, &statfs);
+		if (error)
+			goto out_unmark;
+		/*
+		 * Allow for some extra space for overhead/races. Divide
+		 * first to avoid potential overflow. We know the product
+		 * previously fit into availbytes from dmu_objset_space().
+		 */
+		if (len > statfs.f_bsize / 8 * 7 * statfs.f_bavail) {
```
[the remark this answers is not in the document]
Comment (adilger) [reply module/os/linux/zfs/zpl_file.c:781]: No, the code is correct as-is I think?  This is *decreasing* the calculated free space by 1/8, rather than *increasing* the write size by 1/8 (i.e. not `len * 9 / 8`), to avoid overflowing if a very large `len` is passed.  While `f_bsize` is always going to be at least 512, there is no chance of the /8 resulting in a zero size, but if `f_bavail` = 0 then clearly there isn't enough space in the filesystem for the `fallocate()` to return success.

[the remark this answers is not in the document]
Comment (behlendorf) [reply module/os/linux/zfs/zpl_file.c:781]: My mistake, you're right.

[the remark this answers is not in the document]
Comment (adilger) [reply module/os/linux/zfs/zpl_file.c:781]: In closer review of the code, I realize that this could also have been written in the more normal "* 7 / 8" format since this would be applied to `f_bsize` first and could not overflow the intermediate result due to the maximum `f_bsize` being in the 1MiB range.  I'll fix that in the updated version incorporating your parameter.

**Comment by adamdmoss** [comment]
adilger:basic-fallocate does a good impression of satisfying Steam's pre-allocation and stopping it from doing its own slow big dummy allocation.

File: man/man5/zfs-module-parameters.5 [hunk 4]
```diff
@@ -1658,6 +1658,25 @@ as a percentage of \fBzfs_dirty_data_max\fR.  This should be less than
...
+.na
+\fBzfs_fallocate_reserve_percent\fR (uint)
+.ad
+.RS 12n
+Since ZFS is a copy-on-write filesystem with snapshots, blocks cannot be
+preallocated for a file in order to guarantee that later writes will not
+run out of space.  Instead, fallocate() space preallocation only checks
+that sufficient space is currently available in the pool or the user's
+project quota allocation, and then creates a sparse file of the requested
+size. The requested space is inflated by \fBzfs_fallocate_reserve_percent\fR
```
Comment (ahrens) [inline man/man5/zfs-module-parameters.5:1672]: I would interpret "inflated by" to mean "increased by", i.e. you need requested_size + requested_size * zfs_fallocate_reserve_percent`.  Suggest rewording this:

```suggestion
size. The requested space is multiplied by \fBzfs_fallocate_reserve_percent\fR
```

What happens if this is set between 1 and 99?  We check for less space than was requested?  That might be reasonable if we expect compression (or dedup) to take effect.

File: module/os/linux/zfs/zpl_file.c [hunk 5]
```diff
@@ -743,21 +751,54 @@ zpl_fallocate_common(struct inode *ip, int mode, loff_t offset, loff_t len)
...
+		if (percent == 0) {
+			error = -EOPNOTSUPP;
+			goto out_unmark;
+		}
+
+		/*
+		 * Use zfs_statvfs() instead of dmu_objset_space() since it
+		 * also checks project quota limits, which are relevant here.
+		 */
+		error = zfs_statvfs(ip, &statfs);
```
Comment (ahrens) [inline module/os/linux/zfs/zpl_file.c:785]: I was surprised to find that zfs_statvfs() gets stats related to a specific file (and its project), as opposed to the "vfs" (filesystem).  As it stands, fallocate() / zfs_statvfs() ignores user/group quotas.  Perhaps as a subject of a future PR, we could rename zfs_statvfs to be more accurate, and also have it take into account user/group quotas.

Comment (adilger) [reply module/os/linux/zfs/zpl_file.c:785]: This emulates the behavior of XFS/ext4/others w.r.t. statfs on directories with projectID set on them.  The intent is to set a recursive project ID on a directory tree and treat it like a directory quota/dataset, which is useful for filesystems that don't actually support datasets with their own quota like ZFS.  Changing this for UID/GID quotas might be confusing/surprising for users, maybe not, but would definitely be a big change from all previous history, while the semantics of statfs on directories project quotas has always been this way.

File: man/man5/zfs-module-parameters.5 (hunk 4 above)
Comment (adilger) [reply man/man5/zfs-module-parameters.5]: Yes, I mentioned this in the earlier discussion "In theory it would even be possible to set zpl_fallocate_ratio < 100 to account for compression."  Whether that is sane or not depends on the usage (i.e. it would be bad to assume compression if receiving .mpg files), but at least it is conceivable that this makes sense for some workloads, and would have to actively be changed by the administrator, so I don't think it should be prevented.

Comment (adilger) [reply man/man5/zfs-module-parameters.5]: I didn't realize that this would result in a separate patch rather than being merged into the original commit.  It doesn't make sense IMHO to have a separate one-line patch for this tiny correction, so I've re-pushed with this change merged in.

File: module/os/linux/zfs/zpl_file.c (hunk 5 above)
Comment (ahrens) [reply module/os/linux/zfs/zpl_file.c:785]: Thanks for the context.  Seems like it would be semantically cleaner to just do the more limited change of making fallocate() take into account user/group quotas.
